### ISSUE TITLE
Support selecting cards based on card name

### ISSUE THREAD

**Issue by languitar** [mopidy/mopidy-alsamixer#6]
Numeric indexing of cards is sometimes hard to get persistent across reboots. Please also allow to use card names instead of the index for selecting the device to change.

**Comment by jodal** [comment]
I agree that this would be a big improvement, but I don't immediately see a way to implement it using the [current pyalsaudio API](https://larsimmisch.github.io/pyalsaaudio/):

- We need a `cardindex` or `device` to create the [alsaaudio.Mixer](https://larsimmisch.github.io/pyalsaaudio/libalsaaudio.html#alsaaudio.Mixer).
- `cardindex` is an integer like 0, 1... that maps to a hardware card by name: `hw:0`, `hw:1`...
- `device` is the name of a virtual device, where "default" is the only value I've found working.
- [alsaaudio.cards()](https://larsimmisch.github.io/pyalsaaudio/libalsaaudio.html#alsaaudio.cards) lists the names of all available cards, but the index of a name in the list doesn't necessarily match the `cardindex` we need for `alsaaudio.Mixer()`, as cards may be disabled and thus not be present in the list, offsetting the index for the rest of the cards (see #8 for details). Thus, we cannot rely on the card list returned by `alsaaudio.cards()`.
- We can try to create mixers for `range(len(alsaaudio.cards()))`, an check if we've found the correct one by looking at the name returned from [Mixer.cardname()](https://larsimmisch.github.io/pyalsaaudio/libalsaaudio.html#alsaaudio.Mixer.cardname), but on my system, this simply returns `hw:2`, which doesn't tell me more than this being the card with `cardindex` "2".

If anyone has suggestions how to implement this using the current pyalsaaudio API, I'm all ears.

**Comment by riegaz** [comment]
@jodal could you add an option which uses the currently selected audio device? For example, on my raspberry pi zero w I do have Analog, HDMI and the connected bluetooth speakers. Now I selected the speakers to be the default output. Would be nice to automatically output to the current audio device.

**Comment by robcee** [comment]
Taking a look at this. Not sure if @jodal's comments from 2017 are still valid or not. If I can enumerate properties of audio devices using PyAlsaAudio, we can probably map name to id.

**Comment by robcee** [comment]
just a quick comment, setting up my dev environment to work on this I noticed that we're depending on Gi, which requires Gtk and Cairo and a bunch of other high level libraries that might not be applicable on a headless environment.

It looks like we're only using it for the GstAudio.StreamVolume.convert_volume() method.

Has anyone looked at using one of the numeric libraries (e.g., numpy) for doing the equivalent without the big overhead? Happy to file a follow-up issue to do that conversion.

**Comment by robcee** [comment]
just an update, I have looked at this, and my initial thoughts were to just pull alsaaudio.cards() and do a lookup, but as stated above,

- alsaaudio.cards() lists the names of all available cards, but the index of a name in the list doesn't necessarily match the cardindex we need for alsaaudio.Mixer(), as cards may be disabled and thus not be present in the list, offsetting the index for the rest of the cards (see #8 for details). Thus, we cannot rely on the card list returned by alsaaudio.cards().
- We can try to create mixers for range(len(alsaaudio.cards())), an check if we've found the correct one by looking at the name returned from Mixer.cardname(), but on my system, this simply returns hw:2, which doesn't tell me more than this being the card with cardindex "2".

I think the real fix for this would be to implement an "allcards" method in GstAudio, which, well, doesn't really apply to this issue.

I did manage to get my environment setup for Mopidy development though, so hey, that's something. :)

**Comment by sixtyfive** [comment]
Just stumbled across this googling for how to select by name using `alsaaudio.Mixer()`. This is [now possible](https://larsimmisch.github.io/pyalsaaudio/libalsaaudio.html#mixer-objects) ... thought y'all might want to know.

**Comment by jodal** [comment]
> *This was generated by AI during triage.*

Since v3.0.0 you can set `alsamixer/device` to an ALSA device string, which can select the card by name, for example `device = hw:CARD=Headphones`. Run `aplay -l` to see the card names. Closing as completed.
